core_authentication_service/forms.py

```
@@ -0,0 +1,193 @@
+from django.db.models import QuerySet
+from django.forms import BaseFormSet
+from django.forms import formset_factory
+from django.utils.functional import cached_property
+
+from core_authentication_service import models
+from core_authentication_service.utils import update_form_fields
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+REQUIREMENT_DEFINITION = {
+    "names": ["username", "first_name", "last_name", "nickname"],
+    "picture": ["avatar"]
+}
+
+SECURITY_QUESTION_COUNT = 2
+
+class RegistrationForm(UserCreationForm):
+
+    class Meta:
+        model = get_user_model()
+        fields = [
+            "username", "first_name", "last_name", "email",
+            "nickname", "msisdn", "gender", "birth_date", "country", "avatar"
+        ]
+
+    def __init__(self, security=None, required=[],*args, **kwargs):
+        # Security value is required later in form processes as well.
+        self.security = security
+
+        # Super needed before we can actually update the form.
+        super(RegistrationForm, self).__init__(*args, **kwargs)
+
+        # Set update form update variables, for manipulation as init
+        # progresses.
+        required_fields = []
+        fields_data = {}
+        if required:
+            for field in required:
+                required_fields += REQUIREMENT_DEFINITION.get(field, [field])
+
+        # Security level needs some additions before the form is rendered.
+        if self.security == "high":
+            required_fields += ["email"]
+        else:
+            fields_data = {
+                "password1": {
+                    "attributes": {
+                        "help_text": ""
+                    }
+                }
+            }
+
+        incorrect_fields = set(required_fields) - set(self.fields.keys())
+        for field in incorrect_fields:
+            LOGGER.warning(
+                "Received field to alter that is not on form: %s" % field
+            )
```

Should we not rather raise an exception if there are incorrect fields? These are programmer errors which we cannot allow.

---

The reason I am not keen to just raise an error is, it will most likely later on not be an error on our side. But if we just raise, it will look like our app is failing, from a user perspective. 

Personally I feel that it is not directly an issue with our form and the user should not be effected, but we do need some form of logging to ensure we can point out what the issue was.

---

You are correct. It is not a programmer error because the values are passed in the URL. I missed that bit.